src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -603,6 +607,112 @@ private void EmitArray<T>(Type arrType, IEnumerable<T> items, Action<T, int> ite
+            {
+                for (int i = 0; i < toPopCount; i++)
+                {
+                    EmitOpCode(OpCodes.Pop);
+                }
+
+                return (false, null);
+            }
+            else if (t == typeof(IDictionary))
+            {
+                if (t.IsGenericType)
+                {
+                    return (true, t.GenericTypeArguments[1]);
+                }
+                else
+                {
+                    return (true, typeof(object));
+                }
+            }
+            else if (t.IsArray)
+            {
+                TryEmitIndexingForArray(fcn.FormalArguments[0], fcn.FormalArguments[1], t.GetElementType());
+                return (true,t.GetElementType());
+            }
+            // TODO we may want to bail for IList currently and let 
+            // replication handle it as Ilist is usually a replicated output, and is nested.
+            //Today either we end up emitting IList index opcodes incorrectly for all func return vals,
+            //or we'll call the wrong overload until overload matching is fixed.
+            else if (t == typeof(IList))
+            {
+                if (t.IsGenericType)
+                {
+                    TryEmitIndexingForIList(fcn.FormalArguments[0], fcn.FormalArguments[1],t, t.GenericTypeArguments[0]);
+                    return (true, t.GenericTypeArguments[0]);
+                }
+                else
+                {
+                    TryEmitIndexingForIList(fcn.FormalArguments[0], fcn.FormalArguments[1], t,typeof(object));
+                    return (true,typeof(object));
+                }
+            }
+            EmitILComment("NOT ENOUGH TYPE INFO TO EMIT INDEXING");
+            for (int i = 0; i < toPopCount; i++)
+            {
+                EmitOpCode(OpCodes.Pop);
+            }
+            return (false,null);
+        }
+
+       
+        private bool TryEmitIndexingForIList(AssociativeNode array, AssociativeNode index, Type collectionType, Type listElementType)
+        {
+            var indexT = DfsTraverse(index);
+            var mi = typeof(IList).GetMethod("get_Item", BindingFlags.Instance|BindingFlags.Public);
+            if (collectionType.IsGenericType)
+            {
+                mi = collectionType.GetMethod("get_Item");
+            }
+            EmitOpCode(OpCodes.Callvirt, mi);
+            EmitILComment("INDEX ILST OPERATION END");
+
+            return true;
+        }
+
+        private bool TryEmitIndexingForDictionary(AssociativeNode array, AssociativeNode index, Type collectionType, Type listElementType)
+        {
+           //TODO - Dictionaries in DS will always(almost?) be the DS Dictionary wrapper - which is not a Dictionary
+           //we have to either give it an indexer and call that or make a call to ValueAtKey()
+           //if we think we will encounter regular IDictionary here than we can still implement this.
+
+            return false;
```

Comment:
If we have an `IDictionary `returned from an arbitrary function call and the next AST is something like:
`Builtin.Get.ValueAtIndex(dict, "key");`, I suppose this could be converted to an opcode that directly indexes into an `IDictionary`, assuming we know that `dict `is an `IDictionary`.